v2/pkg/controller/mpi_job_controller_test.go

```diff
@@ -889,7 +890,386 @@ func TestWorkerReady(t *testing.T) {
 	f.run(getKey(mpiJob, t))
 }
 
-func int32Ptr(i int32) *int32 { return &i }
+func TestNewLauncherAndWorker(t *testing.T) {
+	cases := map[string]struct {
+		job          kubeflow.MPIJob
+		workerIndex  int
+		wantLauncher corev1.Pod
+		wantWorker   corev1.Pod
+	}{
+		"defaults": {
+			job: kubeflow.MPIJob{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "foo",
+					Namespace: "bar",
+				},
+				Spec: kubeflow.MPIJobSpec{
+					MPIReplicaSpecs: map[kubeflow.MPIReplicaType]*common.ReplicaSpec{
+						kubeflow.MPIReplicaTypeLauncher: {
+							Template: corev1.PodTemplateSpec{
+								Spec: corev1.PodSpec{
+									Containers: []corev1.Container{{}},
+								},
+							},
+						},
+						kubeflow.MPIReplicaTypeWorker: {
+							Template: corev1.PodTemplateSpec{
+								Spec: corev1.PodSpec{
+									Containers: []corev1.Container{{}},
+								},
+							},
+						},
+					},
+				},
+			},
+			wantLauncher: corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "foo-launcher",
+					Namespace: "bar",
+					Labels: map[string]string{
+						"group-name":   "kubeflow.org",
+						"mpi-job-name": "foo",
+						"mpi-job-role": "launcher",
+					},
+				},
+				Spec: corev1.PodSpec{
+					Hostname:      "foo-launcher",
+					Subdomain:     "foo-worker",
+					RestartPolicy: corev1.RestartPolicyNever,
+					Containers: []corev1.Container{
+						{
+							Env: func() []corev1.EnvVar {
```

Comment:
Can all of this be replaced by "append(ompiEnvVars, nvidiaDisableEnvVars...)?

Reply:
I considered it, but not sure if it will have consequences on the allocation of `ompiEnvVars`

another alternative would be:

```
append(append([]corev1.EnvVar(nil), ompiEnvVars...), nvidiaDisableEnvVars...)
```

WDYT?

Reply:
Yeah, I've just played with append a little and it's a little funny. [1]: "If it has sufficient capacity, the destination is resliced to accommodate the new elements. If it does not, a new underlying array will be allocated."

[1] https://pkg.go.dev/builtin#append

So in this case "append(ompiEnvVars, nvidiaDisableEnvVars...)" will work the same as "append(append(...)", but if you allocate additional memory for the ompiEnvVars slice, it will have a different effect. Examples:

`a := []int{1}
b := append(a[:1], []int{3}...)
a[0] = 2
b[0] = 3
fmt.Printf("%d %d\n", a[0], b[0])` <- this prints "2 3"

`a := []int{1, 2}
b := append(a[:1], []int{3}...)
a[0] = 2
b[0] = 3
fmt.Printf("%d %d\n", a[0], b[0])` <- this prints "3 3"

So append(append([]corev1.EnvVar(nil), ompiEnvVars...), nvidiaDisableEnvVars...) would be better here.

Reply:
Yeah, that will be more predictable. I decided to add a helper method.